### PR TITLE
test: add more test cases

### PR DESCRIPTION
**Pull request by json-derulo**

### CODE REVIEW

**Comment by gund** [comment]
Hey, thanks for the PR!
If this is supposed to cover #25 then the tests should fail but why they are not failing?

**Comment by json-derulo** [comment]
> Hey, thanks for the PR! If this is supposed to cover #25 then the tests should fail but why they are not failing?

Hi! They are not supposed to cover #25. Those are just additional test scenarios which I recognized that they are not covered yet, while trying to implement a fix. The code for covering #25 is the following:

```typescript
getValidTestCase(`
  class Class1 {
    /** @deprecated */ method() {};
  }
  class Class2 extends Class1 {
    method() {};
  }
  new Class2().method();
`),
```
This test case fails currently.

**Review by gund — CHANGES_REQUESTED** [review]
I see, no we don't need to add broken tests =)
I would also ask you to please add to your new test cases a one line comment to each explaining the use-case like in the other cases there.

File: tests/rules/deprecation.test.ts [hunk 1]
```diff
@@ -169,6 +169,38 @@ ruleTester.run('deprecation', rule, {
     getValidTestCase(`
       const component = <Component/>;
     `),
+    getValidTestCase(`
```
Comment (gund) [inline tests/rules/deprecation.test.ts:172]:
```suggestion
    // Method overloads in classes
    getValidTestCase(`
```

File: tests/rules/deprecation.test.ts [hunk 2]
```diff
@@ -169,6 +169,38 @@ ruleTester.run('deprecation', rule, {
...
+
+      method(param: any): void {}
+    }
+    new Class().method('');
+    const obj = new Class();
+    obj.method('');
+  `),
+  getValidTestCase(`
+    class Class {}
+    interface Interface extends Class {
```
Comment (gund) [inline tests/rules/deprecation.test.ts:185]: This does not make much sense as interfaces should not extend classes.
What is the purpose of this?

File: tests/rules/deprecation.test.ts [hunk 3]
```diff
@@ -169,6 +169,38 @@ ruleTester.run('deprecation', rule, {
...
+      method(param: string): void;
+      /** @deprecated */ method(param: number): void;
+
+      method(param: any): void {}
+    }
+    new Class().method('');
+    const obj = new Class();
+    obj.method('');
+  `),
+  getValidTestCase(`
```
Comment (gund) [inline tests/rules/deprecation.test.ts:183]:
```suggestion
  // Method overloads in interfaces
  getValidTestCase(`
```

File: tests/rules/deprecation.test.ts [hunk 4]
```diff
@@ -169,6 +169,38 @@ ruleTester.run('deprecation', rule, {
...
+    interface Interface {}
+    class Class implements Interface {
+      method(param: string): void;
+      /** @deprecated */ method(param: number): void;
+
+      method(param: any): void {}
+    }
+    new Class().method('');
+    const obj = new Class();
+    obj.method('');
```
Comment (gund) [inline tests/rules/deprecation.test.ts:202]: This test case seems to be exactly the same as the first one with extra interface implementation which does not seem like it adds any value to the test case.
Can we remove this one?

File: tests/rules/deprecation.test.ts [hunk 5]
```diff
@@ -459,6 +491,26 @@ ruleTester.run('deprecation', rule, {
       const def4: Interface1 = {};    // ERROR: Interface1
       const component = <Component/>  // ERROR: Component
       `),
+    getInvalidTestCase(`
```
Comment (gund) [inline tests/rules/deprecation.test.ts:494]:
```suggestion
    // Method overloads in interfaces
    getInvalidTestCase(`
```

File: tests/rules/deprecation.test.ts [hunk 6]
```diff
@@ -459,6 +491,26 @@ ruleTester.run('deprecation', rule, {
...
+      interface Interface {
+        method(param: string): void;
+        /** @deprecated */ method(): void;
+        /** @deprecated */ method(param: number): void;
+      }
+      const obj: Interface = { method(args: any) {} };
+      obj.method();  // ERROR: method
+      obj.method(1); // ERROR: method
+    `),
+    getInvalidTestCase(`
```
Comment (gund) [inline tests/rules/deprecation.test.ts:504]:
```suggestion
    // Method overloads in classes
    getInvalidTestCase(`
```

File: tests/rules/deprecation.test.ts [hunk 7]
```diff
@@ -459,6 +491,26 @@ ruleTester.run('deprecation', rule, {
...
       `),
+    getInvalidTestCase(`
+      interface Interface {
+        method(param: string): void;
+        /** @deprecated */ method(): void;
+        /** @deprecated */ method(param: number): void;
+      }
+      const obj: Interface = { method(args: any) {} };
+      obj.method();  // ERROR: method
+      obj.method(1); // ERROR: method
```
Comment (gund) [inline tests/rules/deprecation.test.ts:502]: Please add check that non-deprecated signature is not errored.
```suggestion
      obj.method('valid');  // missed
      obj.method();  // ERROR: method
      obj.method(1); // ERROR: method
```

File: tests/rules/deprecation.test.ts [hunk 8]
```diff
@@ -459,6 +491,26 @@ ruleTester.run('deprecation', rule, {
...
+    `),
+    getInvalidTestCase(`
+      class Class {
+        method(param: string): void;
+        /** @deprecated */ method(): void;
+        /** @deprecated */ method(param: number): void;
+      }
+      const obj = new Class();
+      obj.method();  // ERROR: method
+      obj.method(1); // ERROR: method
```
Comment (gund) [inline tests/rules/deprecation.test.ts:512]: Please add check that non-deprecated signature is not errored.
```suggestion
      obj.method('valid');  // missed
      obj.method();  // ERROR: method
      obj.method(1); // ERROR: method
```

File: tests/rules/deprecation.test.ts (hunk 2 above)
Comment (json-derulo) [reply tests/rules/deprecation.test.ts:185]: In a previous version of the TypeScript handbook, this exact technique was mentioned, see https://www.typescriptlang.org/docs/handbook/classes.html#using-a-class-as-an-interface. It's not mentioned anymore in the current version of the handbook, however it is still used in real world.

**Review by gund — APPROVED** [review]
Thanks for the fixes! Looks good!

**Pull request merged**